proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,1058 @@
+    import splice 1 A
+    -- Can be used at level -2
+    import splice 2 A
+
+Practically, by far the most common situation is a single level of splices, so in the interests
+of reducing complexity we do not propose supporting this at present.
+
+
+Inferring splice/quote imports based on usage
+#############################################
+
+Since our proposed approach has the type-checker verify that usage of ``splice``
+or ``quote`` imports is correct, it may be possible in principle to infer where
+``splice`` or ``quote`` keywords are needed, based on usage inside a module.
+However, this would compromise the principle that the build system can discover
+the dependencies for a module just by looking at the import list in the module
+header. Achieving the performance benefits of our proposed approach would
+involve significant technical complexity (as the compiler would need to
+partially type-check a module, then suspend compilation of that module while it
+compiles those of its dependencies determined to be required for further
+type-checking).
+
+Given that the ``splice`` and ``quote`` annotations are useful for human readers
+understanding how code is staged, it seems worthwhile to make them explicit.
+
+Of course, nothing prevents development of a tool that helps users insert
+``splice`` and ``quote`` annotations into their modules as part of a migration
+to using ``ExplicitLevelImports``.
+
+
+Syntactic alternatives
+######################
+
+* Using a pragma rather than a syntactic modifier would fit in better with
+  how ``SOURCE`` imports work and make writing backwards compatible code easier::
+
+    import {-# SPLICE #-} B
+
+* There are several proposals for the syntax of splice imports. Some have objected
+  that the ``import splice`` suggestion is ungrammatical, unlike ``import qualified`` or
+  ``import hiding``.
+
+  One possible alternative is ``$(import Foo)`` to represent a splice import, but this
+  syntax clashes with the existing syntax for declaration splices and significantly
+  changes the structure of the import syntax.
+
+  Another alternative suggested was ``import for splice``, which restores the
+  grammatical nature of the import.
+
+
+Other alternatives
+##################
+
+* The extension could apply only to "home" modules (those from the package being compiled), because the primary benefits of
+  splice imports are when using GHC's ``--make`` mode. As the proposal stands,
+  for uniformity, any module used inside a top-level splice must be marked as
+  a splice import, even if it's from an external package.
+
+* Since ``ExplicitLevelImports`` is essentially useless when
+  ``TemplateHaskell`` is disabled, we could have ``ExplicitLevelImports`` imply
+  ``TemplateHaskell``.  There is at least one case where this would be harmful:
+  users may which to enable ``ExplicitLevelImports`` globally for their
+  project, but only carefully enable ``TemplateHaskell`` for a small number of
+  modules.
```

Comment:
In what way is this harmful? Beyond hurting the users' feeling, that is. Are there practical consequences?

Reply:
If you enable `TemplateHaskell` in a module, then this indicates to GHC that you require the modules dependencies to be compiled before the module is able to be compiled (in case that any dependency is needed to execute a top-level splice).